src/lib/generateAhoXml.ts

```diff
@@ -0,0 +1,362 @@
+import { format } from "date-fns"
+import { XMLBuilder } from "fast-xml-parser"
+import type {
+  Adj,
+  AhoParsedXml,
+  Br7Case,
+  Br7Duration,
+  Br7Hearing,
+  Br7Offence,
+  Br7OffenceReason,
+  Br7OrganisationUnit,
+  Br7Result,
+  Br7Urgent,
+  Cxe01,
+  DISList
+} from "src/types/AhoParsedXml"
+import type {
+  AnnotatedHearingOutcome,
+  Case,
+  Duration,
+  Hearing,
+  Offence,
+  OffenceReason,
+  OrganisationUnit,
+  Result,
+  Urgent
+} from "src/types/AnnotatedHearingOutcome"
+import type { PNCDisposal, PncQueryResult } from "src/types/PncQueryResult"
+import {
+  lookupDefendantPresentAtHearingByCjsCode,
+  lookupModeOfTrialReasonByCjsCode,
+  lookupOffenceCategoryByCjsCode,
+  lookupOffenceDateCodeByCjsCode,
+  lookupSummonsCodeByCjsCode,
+  lookupVerdictByCjsCode
+} from "src/use-cases/dataLookup"
+
+const mapAhoOrgUnitToXml = (orgUnit: OrganisationUnit): Br7OrganisationUnit => ({
+  "ds:TopLevelCode": orgUnit.TopLevelCode,
+  "ds:SecondLevelCode": orgUnit.SecondLevelCode,
+  "ds:ThirdLevelCode": orgUnit.ThirdLevelCode,
+  "ds:BottomLevelCode": orgUnit.BottomLevelCode,
+  "ds:OrganisationUnitCode": orgUnit.OrganisationUnitCode,
+  "@_SchemaVersion": "2.0"
+})
+
+const mapAhoUrgentToXml = (urgent: Urgent): Br7Urgent => ({
+  "br7:urgent": { "#text": urgent.urgent ? "Y" : "N", "@_Literal": urgent.urgent ? "Yes" : "No" },
+  "br7:urgency": urgent.urgency
+})
+
+const mapAhoDuration = (duration: Duration[]): Br7Duration[] =>
+  duration.map((d) => ({
+    "ds:DurationType": d.DurationType,
+    "ds:DurationUnit": d.DurationUnit,
+    "ds:DurationLength": d.DurationLength
+  }))
+
+const mapAhoResultsToXml = (results: Result[]): Br7Result[] =>
+  results.map((result) => ({
+    "ds:CJSresultCode": result.CJSresultCode,
+    "ds:SourceOrganisation": mapAhoOrgUnitToXml(result.SourceOrganisation),
+    "ds:CourtType": result.CourtType,
+    "ds:ResultHearingType": { "#text": result.ResultHearingType, "@_Literal": "Other" },
+    "ds:ResultHearingDate": result.ResultHearingDate ? format(result.ResultHearingDate, "yyyy-MM-dd") : "",
```

Comment:
suggestion: should we leave the XML fields out when things don't exist on the AHO, instead of having fields present with the empty string value? It will make the code handing this XML more complex when it has to check for fields that exist but aren't empty, and imo it would be more of a "true" representation to have these fields be optional in the types and not have the fields present in the XML when they're not in the AHO, but there might be some context on this I'm missing

Reply:
Good point, have done this now. I left along any keys that map to attributes in the xml (anything that looks like `@_thing`) as these seem to want to be empty strings.